modify deepsleep to allow wakeup on X1 or X18 on a STM32F7 (SF2/3/6W)

irsla opened the pull request:
…F2/6W

example
machine.deepsleep([time_ms | 0 for infinity], [WKUP_PINS_X1/X18], [WKUP_PINS_X1/X18_RISING/FALLING])

machine.deepsleep()
machine.deepsleep(20000)
machine.deepsleep(10000, machine.WKUP_X1, machine.WKUP_X1_RISING)
machine.deepsleep(0, machine.WKUP_X1, machine.WKUP_X1_RISING)
machine.deepsleep(0, machine.WKUP_X1, machine.WKUP_X1_FALLING)

Prior to this commit only the timer wakeup would work for a deepsleep.
This was the case because for the STM32F7 all wakeup pins were cleared out before
entering in deepsleep

this commit is a work in progress for technical discussions.

irsla committed:
modify deepsleep to allow wakeup on X1 or X18 (to be tested !) on a SF2/6W

example
machine.deepsleep([time_ms | 0 for infinity], [WKUP_PINS_X1/X18], [WKUP_PINS_X1/X18_RISING/FALLING])

machine.deepsleep()
machine.deepsleep(20000)
machine.deepsleep(10000, machine.WKUP_X1, machine.WKUP_X1_RISING)
machine.deepsleep(0, machine.WKUP_X1, machine.WKUP_X1_RISING)
machine.deepsleep(0, machine.WKUP_X1, machine.WKUP_X1_FALLING)

Prior to this commit only the timer wakeup would work for a deepsleep.
This was the case because for the STM32F7 all wakeup pins were cleared out before
entering in deepsleep

this commit is a work in progress for technical discussions.

peterhinch commented:
This is STM specific. If a C solution is to be used, **pyb** would be preferable to **machine** which is cross-platform. In my view a Python solution is preferable. Speed is not an issue, and adding a specialist feature to the firmware could be seen as needless bloat.

I have a long standing solution for Pyboard 1.x [here](https://github.com/peterhinch/micropython-micropower). This is written in Python via the **stm** module. I am currently porting this to Pyboard D. The pin and RTC wakeups are working and I expect to push an update in the coming days.

An official solution would be ideal, but (in my view) best written in Python.

irsla commented:
Ok now we have a duplicate for this issue, which is good, it means that it will be solved one way or another.

So now we need to discuss the solutions.

Solution 1 (time based): PR 6482 (this one):
   - provide full control over X1 and X18 wake up pins in the machine/pyb module in an "easy way"
           - full control:
                      - set wake up pin
                      - set trigger on falling/raising wake up 
                      - get the reset_cause
Solution 2 (time based): PR 6494
   - handle full control over X1 and X18 to the user 
           - full control:
                       - set wake up pin in python
                      - set trigger on falling/raising wake up in python
                      - get the reset_cause in python

Both solution have advantages and drawbacks.

Drawbacks:
   Solution 1: 
          a lot of C code as we will need to break the calling of machine_deepsleep from pyb_standby in order to control the pins and will have to duplicate the reset_cause and have a fall-through machine_reset_cause if pyb_reset_cause is not X1/X18

  Solution 2:
        the complexity is moved to the user/developer 

How can we have this discussion ?

Duplicate: https://github.com/micropython/micropython/pull/6494

peterhinch commented:
The latest release of [my micropower repo](https://github.com/peterhinch/micropython-micropower) requires only this small change to the firmware. It supports the Pyboard D in a way that has been in use for some time on the Pyboard 1.x. User code will only need minor changes to migrate to the D series. Further this very simple PR doesn't preclude supporting micropower functions in C. It merely provides to the D series functionality available on the Pyboard 1.x. In my opinion it is a bug fix.

I see no conflict between the different approaches. My solution exists now and (with this bug fix) works on the D series. If there is a case for a C solution I see no reason why this cannot be developed independently.

irsla committed:
Proposal #2

this includes patch for PR6494 and allows both patch to co-exists

Users can then choose between 2 ways
   machine.deepsleep([ms])
   pyb.standby([ms], [wakeup Pins X1|X18], [wakeup trigger FALLING|RISING])

-- tested 1.13
machine.deepsleep(2000) # OK
machine.deepsleep(5000) # OK
machine.deepsleep(10000) # OK
machine.deepsleep(20000) # OK
machine.deepsleep() # OK waited 2 minutes

-- tested new firmware
machine.deepsleep(2000) # OK
machine.deepsleep(5000) # OK
machine.deepsleep(10000) # OK
machine.deepsleep(20000) # OK
machine.deepsleep() # OK waited 2 minutes

pyb.standby(0) # OK waited for 2 minutes
pyb.standby(2000) # OK
pyb.standby(5000) # OK
pyb.standby(10000) # OK
pyb.standby(20000) # OK
pyb.standby() # OK waited for 2 minutes

pyb.standby(0, pyb.WKUP_X1, pyb.WKUP_X1_FALLING) # OK wake up on interrupt within 10 sec
pyb.standby(2000, pyb.WKUP_X1, pyb.WKUP_X1_FALLING) # OK wakes up after 2sec
pyb.standby(5000, pyb.WKUP_X1, pyb.WKUP_X1_FALLING) # OK wakes up after 5 sec
pyb.standby(10000, pyb.WKUP_X1, pyb.WKUP_X1_FALLING) # OK wakes up after 10 sec and wakes up on interrupt within 10 sec
pyb.standby(20000, pyb.WKUP_X1, pyb.WKUP_X1_FALLING) # OK wakes up after 20 sec  and wakes up on interrupt within 20 sec

pyb.standby(0, pyb.WKUP_X18, pyb.WKUP_X18_FALLING) # OK wake up on interrupt within 10 sec
pyb.standby(2000, pyb.WKUP_X18, pyb.WKUP_X18_FALLING) # OK wakes up after 2sec
pyb.standby(5000, pyb.WKUP_X18, pyb.WKUP_X18_FALLING) # OK wakes up after 5 sec
pyb.standby(10000, pyb.WKUP_X18, pyb.WKUP_X18_FALLING) # OK wakes up after 10 sec and wakes up on interrupt within 10 sec
pyb.standby(20000, pyb.WKUP_X18, pyb.WKUP_X18_FALLING) # OK wakes up after 20 sec  and wakes up on interrupt within 20 sec

-- TO TEST
pyb.standby(0, pyb.WKUP_X1 | pyb.WKUP_X18, pyb.WKUP_X1_FALLING | pyb.WKUP_X18_FALLING) #
machine.deepsleep with upower.py

irsla commented:
So here is V2, where I had to cut the link between pyb.standby and machine.deepsleep to keep PR6494

I would love to have the opinion of a maintainer, so that I know if there's any good in pursuing into this direction

projectgus commented:
This is an automated heads-up that we've just merged a Pull Request
that removes the STATIC macro from MicroPython's C API.

See https://github.com/micropython/micropython/pull/13763

A search suggests this PR might apply the STATIC macro to some C code. If it
does, then next time you rebase the PR (or merge from master) then you should
please replace all the `STATIC` keywords with `static`.

Although this is an automated message, feel free to @-reply to me directly if
you have any questions about this.
